contracts/Vesting.sol

```diff
@@ -0,0 +1,63 @@
+pragma solidity ^0.8.0;
+
+import "@openzeppelin/contracts/access/Ownable.sol";
+import "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import "hardhat/console.sol";
+
+contract Vesting is Ownable {
+    struct Vest {
+        uint256 amount;
+        uint256 timestamp;
+        uint256 collected;
+    }
+
+    mapping(address => Vest) private _vests;
+
+    uint constant _vestingPeriod = 60*60*24*30; // 1 month in seconds
+
+    IERC20 private _token;
+
+    constructor(IERC20 token) {
+        _token = token;
+    }
+
+    function vest(address receiver, uint amount) public onlyOwner {
+        _vests[receiver] = Vest(amount, block.timestamp, 0);
+    }
+
+    function claim() public returns(uint) {
+        uint claimAmount = getCurrentVestAmountForAddress(msg.sender);
+
+        _token.transfer(msg.sender, claimAmount);
```

Comment:
poczytaj o check effects interactions pattern: https://fravoll.github.io/solidity-patterns/checks_effects_interactions.html

Reply:
to jeszcze zastosuj w praktyce ;) zmiana stanu powinna nastąpić przed interackcją z zewnętrznym kontraktem (przenieś `_token.transfer(msg.sender, claimAmount);` jedną operację niżej)